### PR TITLE
vtgate: accept PUT requests to /debug/health

### PR DESCRIPTION
**Pull request by sjmudd**
This is to allow us to change the readiness of the process.
See: https://github.com/vitessio/vitess/issues/4072

Not sure if this is good enough but it looks like it might do what is needed.

Signed-off-by: Simon J Mudd <sjmudd@pobox.com>

### CODE REVIEW

**Comment by sjmudd** [comment]
Note: this is just an idea. not sure if it makes sense or needs to be handled differently.

**Comment by sjmudd** [comment]
Also it may be easier to change the `string` into `[]byte` which avoids unnecessary conversions and simplifies things a bit further.

**Comment by sjmudd** [comment]
Actually this isn't right. `IsHealthy()` does not yet take into account `currentHealth` which it should do.

**Comment by sougou** [comment]
Sorry about coming in late on this one. I had to find time to dig into the server code before I could comment.

Basically, there's already a lameduck mechanism supported by the server framework. It works as follows:
* You set a lameduck period using this command line option: https://github.com/vitessio/vitess/blob/master/go/vt/servenv/servenv.go#L58.
* You can also register a function to be called when the process enters lameduck mode (OnTerm). It looks like there's already one registered for the mysql protocol: https://github.com/vitessio/vitess/blob/master/go/vt/vtgate/plugin_mysql_server.go#L289.
* Once this is set, you can issue a `SIGTERM`, `SIGUSR1` or `SIGKILL` to the process.
* When the process receives this signal, it will enter a lameduck period. When entering, it will fire the OnTerm functions.
* After the lameduck period ends, the process willl exit.

In other words, all you have to do is set vtgate as unhealthy in the `shutdownMysqlProtocolAndDrain` function. And then, whenever you want to terminate a vtgate, send it a `SIGKILL`.

Let me know if this makes sense.

**Comment by sjmudd** [comment]
Hi @sougou, thanks for the detailed explanation.
What you say makes sense but I do not think it quite does what I want.
* I want monitoring systems to have a way to recognise if they should not send requests to vtgate.  Currently this is not handled by the output of `/debug/health`.  _Aliveness_ and _Readiness_ are 2 different things and in this case I would like a way to monitor _readiness_.
* Making vtgate "signal" it is not ready to receive requests may or may not be related to me shutting down vtgate. e.g. it could also be used for testing new code for a short period of time.  So I would like to not force this readiness state to the process shutdown, though most of the time it would be used for this purpose.
* The signals can only be sent from the localhost: it would be convenient for me to be able to trigger the change in state remotely, which is what I was suggesting with the PUT request to `/debug/health` though we can use another URL if that's considered cleaner/better.  Perhaps you do not consider this a good idea for security (remote access) reasons?  I had originally considered using a signal handler for making the process _ready_ or _not_ready_ but then considered PUTting to a URL might be more flexible (assuming suitable access controls are present).

This is partly why I chose to the route I had, but clearly I hadn't considered `shutdownMysqlProtocolAndDrain`, lame-duck mode or the addition of a hook to be called when the process stops being ready. That also needs addressing.

**Comment by sougou** [comment]
This makes sense. One thing we could do is add ACL checks at the beginning, like here: https://github.com/vitessio/vitess/blob/master/go/vt/vtgate/executor.go#L1295

But you'll need to plug an authorization system to be able to enforce ACL. YouTube does use such a plugin internally.

I want to think some more about this to see if there's a more generic approach that would work for all binaries. If I can't come up with something better, we'll go with this.

**Comment by sougou** [comment]
@sjmudd do you still need this feature? If so, let's create an issue instead, and then we can implement the generalized approach for all vitess binaries.

**Comment by sjmudd** [comment]
Hi. right now we are not using Vitess so will close this PR.
